provider/waiter.go

```diff
@@ -0,0 +1,160 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/go-cty/cty"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+)
+
+// Waiter is a simple interface to implement a blocking wait operation
+type Waiter interface {
+	Wait(context.Context) error
+}
+
+// NewResourceWaiter constructs an appropriate Waiter using the supplied waitForBlock configuration
+func NewResourceWaiter(resource dynamic.ResourceInterface, resourceName string, waitForBlock cty.Value) (Waiter, error) {
+	fields := waitForBlock.GetAttr("fields")
+
+	if !fields.IsNull() || fields.IsKnown() {
+		if !fields.Type().IsMapType() {
+			return nil, fmt.Errorf(`"fields" should be a map of strings`)
+		}
+
+		vm := fields.AsValueMap()
+		matchers := []FieldMatcher{}
+		for k, v := range vm {
+			expr := v.AsString()
+			var re *regexp.Regexp
+			if expr == "*" {
+				// NOTE this is just a shorthand so the user doesn't have to
```

Comment:
I'm not sure about this - but I also don't want to burden the user if they just want to wait for any value.